### ISSUE TITLE
'zone' is not a valid value for kind because exceeded maximum number of │ zone rulesets for phase http_ratelimit (20217)

### ISSUE THREAD

**Issue by ratesangelo** [cloudflare/terraform-provider-cloudflare#3444]
### Confirmation

- [X] This is a bug with an existing resource and is not a feature request or enhancement. Feature requests should be submitted with Cloudflare Support or your account team.
- [X] I have searched the issue tracker and my issue isn't already found.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

    cloudflare = {
      source  = "cloudflare/cloudflare"
      version = " 4.27.0"
}

### Affected resource(s)

Zone level

### Terraform configuration files

```hcl
resource "cloudflare_ruleset" "rate_limiting_rules_domain_dev" {
  description = "apply HTTP rate limiting for a route partner-with-us"
  name        = "Requests from ips that exceed requests per minute"
  kind        = "zone"
  phase       = "http_ratelimit"
  zone_id     = cloudflare_zone.reference_to_zone.id
  rules {
    action      = "managed_challenge"
    description = "Protect web form submissions"
    enabled     = true
    expression  = "(http.request.uri.path contains \"partner-with-us\")"
    ratelimit {
      characteristics     = ["cf.colo.id", "ip.src"]
      period              = 60
      requests_per_period = 2
    }
  }
}
```

### Link to debug output

https://gist.github.com/ratesangelo/7dd796e9ac6907c7bc8981ddfd61ffe7

### Expected output

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
Job succeeded

### Actual output

'zone' is not a valid value for kind because exceeded maximum number of
│ zone rulesets for phase http_ratelimit (20217)

and 

│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to
│ cloudflare_ruleset.rate_limiting_rules_lowestrates_dev, provider
│ "provider[\"registry.terraform.io/cloudflare/cloudflare\"]" produced an
│ unexpected new value: .rules[0].ratelimit[0].mitigation_timeout: was null,
│ but now cty.NumberIntVal(0).
│ 
│ This is a bug in the provider, which should be reported in the provider's
│ own issue tracker.

### Steps to reproduce

Context: We aim to migrate all our existing rules in the Cloudflare UI to Terraform since some are already in Terraform code. However, we are struggling with the implementation of the "cloudflare_ruleset" resource.
1) WAF-> Rate Limiting
2) Using CF_Terraforming we got the right syntax as Terraform provider and Cloudflare documentation are not matching.
3) After running the code in our pipeline we got the following error:

A similar configuration with rules already exists and overwriting will have
│ unintended consequences. If you are migrating from the Dashboard, you will
│ need to first import the existing rules using cf-terraforming. You can find
│ details about how to do this at

As the rule exist in UI and Terraform cannot overrun it. 
4) Deleting the rate limiting WAF rule from UI
5) Running the pipeline again with the code from cf_terraforming and getting the following error:

 'zone' is not a valid value for kind because exceeded maximum number of
│ zone rulesets for phase http_ratelimit (20217)

6) googling the error as clearly deleting rule from UI dont really update the backend API 
using API call to list all rulesets in a zone: https://developers.cloudflare.com/api/operations/listZoneRulesets
to find the zone id of the "ghosted rate limiting rule"
7) Deleting the rule via API call: 
https://developers.cloudflare.com/api/operations/deleteZoneRuleset
8) Running the pipeline again with the cf_terraforming code to get :
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to
│ cloudflare_ruleset.rate_limiting_rules_lowestrates_dev, provider
│ "provider[\"registry.terraform.io/cloudflare/cloudflare\"]" produced an
│ unexpected new value: .rules[0].ratelimit[0].mitigation_timeout: was null,
│ but now cty.NumberIntVal(0).
│ 
│ This is a bug in the provider, which should be reported in the provider's
│ own issue tracker.

checking the UI the rule got created! but it terrafrom run failed run failed

Please, Cloudflare Terraform provider team, address the numerous bugs present. It is not acceptable to dismiss these issues as "by design" when significant financial investment is involved.

**Comment by jacobbednarz** [comment]
the error message here informs you of the next steps needed.

> A similar configuration with rules already exists and overwriting will have
> unintended consequences. If you are migrating from the Dashboard, you will
> need to first import the existing rules using cf-terraforming. You can find
> details about how to do this at https://developers.cloudflare.com/terraform/additional-configurations/ddos-managed-rulesets/#optional-delete-existing-rulesets-to-start-from-scratch

if you need to migrate, you need to look at using a tool like [cf-terraforming](https://github.com/cloudflare/cf-terraforming)

**Comment by ratesangelo** [comment]
@jacobbednarz did you read my note:

"Using CF_Terraforming we got the right syntax as Terraform provider and Cloudflare documentation are not matching." 

I need to change 21 zones, some for production and some for development. Given the robustness of the  resource cloudflare_ruleset, initiating such a large-scale change from scratch poses a risk.

**Comment by jacobbednarz** [comment]
you said you got the right syntax? i'm not sure what else you're looking for here. if you're having an issue with the cf-terraforming tool, you can raise an issue in that repository.

**Comment by ratesangelo** [comment]
cf_terraforming is a component of Cloudflare. I just reviewed the repository and noticed your comments there too. At the end it all connected

**Comment by solomonjames** [comment]
> How is it possible that using the UI it works and using the CF_terrafroming it is failing. Either the provider or the cf_terrafroming to its under the same company

Currently having a similar problem. The terraform module gives an error, but I can do the exact same thing in the UI and it works.

**Comment by Eelviny** [comment]
Would love to have this looked at again. I ran into this issue and ended up here, which didn't really tell me how to resolve it. The situation is that if you create some rules from the webUI, even if you delete the rules, the empty "ruleset" object still remains behind. This makes Terraform fail with the above-mentioned error.

To solve this, I opened up F12 developer mode from within the UI, and loaded up the rules page. You'll see it attempts to query for each type of ruleset, and it will return an ID for each one. Then, do a `terraform import` to import that ruleset object, and TF will then be able to modify it.

While this may be obvious from the API, I can imagine I'm not the only user looking to move from the Web UI to Terraform. I would expect that the terraform provider recognise that the object exists and takes it over automatically, if it is empty.
